File: ProjectEuler/prob9_all/prob9.js
// A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,
// a^2 + b^2 = c^2
// For example, 3^2 + 4^2 = 9 + 16 = 25 = 5^2.
// There exists exactly one Pythagorean triplet for which a + b + c = 1000.
// Find the product abc.


// https://en.wikipedia.org/wiki/Formulas_for_generating_Pythagorean_triples
// http://www.mathsisfun.com/numbers/pythagorean-triples.html

var a, b, c, tripletTotal = 1000;
var c_min = Math.ceil((Math.sqrt(2) - 1) * tripletTotal), c_max = tripletTotal/2;  // (sqrt(2) - 1) equals (c / (a + b + c)) in an isosceles right triangle 

// Find all numbers in [c_min, c_max) that can be reached via below methods
/* Dickson method
if (r^2 = 2st)
	then (a = r + s), (b = r + t), (c = r + s + t)
*/
/* MathisFun method
a = n^2 - m^2
b = 2nm
c = n^2 + m^2
*/

if (true)
{
	var MathIsFun = new MathIsFun_technique();
	MathIsFun.popSquares();
	alert(MathIsFun.squares);
} else if (false)
{
}


function MathIsFun_technique ()
{
	this.squares = [];
	this.popSquares = function ()
	{
		var i, j;
		for (i = 1, j = Math.floor(Math.sqrt(c_max)); i <= j; i++)
		{
			// this.squares.push.call(this.caller, Math.pow(i, 2)); // context is wrong
			this.squares.push(Math.pow(i, 2));
		}
	};	
}

function LeonardDickson_technique ()
{
}

File: ProjectEuler/prob9_all/prob9.c
// A Pythagorean triplet is a set of three natural numbers, a < b < c, for which
// a^2 + b^2 = c^2
// For example, 3^2 + 4^2 = 9 + 16 = 25 = 5^2.
// There exists exactly one Pythagorean triplet for which a + b + c = 1000. Find the product abc.

#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <math.h>

bool validate_triple(int, int, int);

int total = 1000;

int main(int argc, char *argv[]) {
	for (int c = 997; c > 0; c -= 4) {
		for (int b = c-1; b > 0; b--) {
			for (int a = b-1; a > 0; a--) {
				if (validate_triple(a, b, c)) {
					printf("Found triple %d %d %d.  It's product is %d.\n", a, b, c, a*b*c);
					return 0;
				}
			}
		}
	}
}

bool validate_triple(int a, int b, int c) {
	bool is_triple = (pow(a, 2) + pow(b, 2) == pow(c, 2));
	bool is_total = (a+b+c == total);

	return is_triple && is_total;
}
